### PR TITLE
Route replay with a smarter driver

### PR DESCRIPTION
**Pull request by kmadsen**
## Description

Resolves https://github.com/mapbox/mapbox-navigation-android/issues/2717

Documentation https://docs.google.com/document/d/1_EX50X_Eu1TVwcX_a-ZK7lRHFQD1HV4S0a7eVqY4ZQs/edit#heading=h.4bzd21t6zoy8

Breaking this PR down into smaller pieces
- [x] ReplayRouteSmoother https://github.com/mapbox/mapbox-navigation-android/pull/2893
- [x] ReplayRouteInterpolator https://github.com/mapbox/mapbox-navigation-android/pull/2908
- [ ] ReplayRouteDriver
- [ ] ReplayRouteMapper
- [ ] ReplayProgressObserver

Route replay has had various issues because it has a constant speed. I think we can fix it! Pretty excited to see that this is mostly working now. So I'm opening up the draft. 

Remaining issues on this are not very easy. The one I'm working on now, is making the simulator reach a position at a specified time and speed. 

Please include a brief summary of the change and which issue is fixed (e.g., fixes [#issue](link))

- [ ] I have added any issue links
- [x] I have added all related labels (`bug`, `feature`, `new API(s)`, `SEMVER`, etc.)
- [x] I have added the appropriate milestone and project boards

## Screenshots or Gifs

| u-turns | curved roads |
| -- | -- |
| output | output |

| stop at waypoints | fast forward and pause |
| -- | -- |
| output | output |

| stop at u-turn waypoints | place holder |
| -- | -- |
| output | place holder |

## Testing

Please describe the manual tests that you ran to verify your changes

- [ ] I have tested locally (including `SNAPSHOT` upstream dependencies if needed) through testapp/demo app and run all activities to avoid regressions
- [x] I have tested via a test drive, or a simulation/mock location app
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

## Checklist

- [x] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have updated the `CHANGELOG` including this PR

### CODE REVIEW

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/history/ReplayRouteDriver.kt [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+            if (index + 3 < gpsLocations.lastIndex) {
+                val fromPoint = Point.fromLngLat(location.longitude, location.latitude)
+                val toPoint = Point.fromLngLat(gpsLocations[index + 3].longitude, gpsLocations[index + 3].latitude)
+                bearing = TurfMeasurement.bearing(fromPoint, toPoint)
+            }
+            location.bearing = bearing.toFloat()
+        }
+    }
+
+    private fun driveSegment(startTime: Int, startSpeed: Int, endSpeed: Int, distance: Double): DriveSegment {
```
Comment (kmadsen) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/history/ReplayRouteDriver.kt:106]: Here is a jupyter notebook used to figure this out
https://github.com/kmadsen/android/blob/master/positioning/interpolation/Intepolate-points.pdf

File: libnavigation-core/src/test/java/com/mapbox/navigation/core/replay/route2/ReplayRouteSmootherTest.kt [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+    fun `should keep points for wide freeway turns`() {
+        val routeSmoother = ReplayRouteSmoother()
+        val geometry = """u{l~fA|kenhF`]vChRl@hRL`Wk@fN_@~HMnILfINvI\dJjAdJzAvIxBnIhC~HvCfIvDfIrEpHrF~HbG~H`H~I~IrJhLvIfNd@|@nDrF~CrFfDpG~CrGpC`GxGtPnCpGhCpGnCrGdj@xjAxLdYhCrGvC`HhCpGfC`HhC`HhC`HxB`H`C`HxBnHpB`HhB`IbB`HzK|gAt@tOd@rP?vN?bp@WpRsAzi@?hCMxa@"""
+        val coordinates = LineString.fromPolyline(geometry, 6).coordinates()
+
+        val smoothedRoutes = routeSmoother.smoothRoute(coordinates, 3.0)
+
+        // Note that this should be 15. Creating a lower bound as we try to improve
+        // the solution for tracking wide turns
+        assertTrue("${smoothedRoutes.size} == 20", smoothedRoutes.size >= 9)
```
Comment (kmadsen) [inline libnavigation-core/src/test/java/com/mapbox/navigation/core/replay/route2/ReplayRouteSmootherTest.kt:149]: | long version | long version clipped to see issue |
| -- | --- |
| output | output |

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/route2/ReplayRouteInterpolator.kt [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+        const val maxAcceleration = 4.0
+        const val minAcceleration = -3.0
+    }
+
+    private val routeSmoother = ReplayRouteSmoother()
+
+    /**
+     *
+     */
+    fun routeSpeedProfile(directionsRoute: DirectionsRoute): List<ReplayRouteLocation> {
```
Comment (kmadsen) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/route2/ReplayRouteInterpolator.kt:39]: still looking to give this some tests and ❤️

File: libnavigation-core/src/test/java/com/mapbox/navigation/core/replay/route2/ReplayRouteSmootherTest.kt (hunk 2 above)
Comment (kmadsen) [reply libnavigation-core/src/test/java/com/mapbox/navigation/core/replay/route2/ReplayRouteSmootherTest.kt:149]: Yeah curvy roads. so far it's built for straight roads 🤔. at least the pieces are all in place to fix this, but it'll take a day or two to build

output

Comment (kmadsen) [reply libnavigation-core/src/test/java/com/mapbox/navigation/core/replay/route2/ReplayRouteSmootherTest.kt:149]: Ok here is the improvement. I'm still looking into the speed stalls (grabs old calculus book)

output

**Comment by Nawelbn** [comment]
Hello, Is there a way to update ReplayRouteLocation's speed during navigation?

**Comment by kmadsen** [comment]
> Hello, Is there a way to update ReplayRouteLocation's speed during navigation?

HI @Nawelbn, do you mean like this?
output

**Comment by kmadsen** [comment]
Finally able to test some multiple stops with the new ArrivalController and ArrivalObserver

output

**Comment by Nawelbn** [comment]
@kmadsen yes, exactly what I was meaning, thanks a lot.  Still have a problem using snapshots, this is the error I get :

Could not find com.mapbox.navigator:mapbox-navigation-native:11.0.2.
Required by:
    project :app > com.mapbox.mapboxsdk:mapbox-android-navigation:0.43.0-SNAPSHOT:20200502.024018-446

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/route2/ReplayRouteMapper.kt [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+import com.mapbox.api.directions.v5.models.LegStep
+import com.mapbox.api.directions.v5.models.RouteLeg
+import com.mapbox.geojson.utils.PolylineUtils
+import com.mapbox.navigation.core.replay.history.ReplayEventBase
+import com.mapbox.navigation.core.replay.history.ReplayEventLocation
+import com.mapbox.navigation.core.replay.history.ReplayEventUpdateLocation
+import com.mapbox.navigation.core.replay.history.ReplayHistoryPlayer
+
+/**
+ * This class converts a directions rout into events that can be
```
Comment (Guardiola31337) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/route2/ReplayRouteMapper.kt:17]:
```suggestion
 * This class converts a [DirectionsRoute] into events that can be
```

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/route2/ReplayRouteMapper.kt [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+     * Options that allow you to control the driver and car behavior
+     */
+    var options: ReplayRouteOptions = ReplayRouteOptions.Builder().build()
+) {
+
+    private val replayRouteDriver = ReplayRouteDriver()
+
+    /**
+     * Take a [DirectionsRoute] and map it to events that can be replayed by the [ReplayHistoryPlayer].
+     * Uses the directions api [DirectionsRoute.geometry] to calculate the speed
```
Comment (Guardiola31337) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/route2/ReplayRouteMapper.kt:31]:
```suggestion
     * Uses the Directions API [DirectionsRoute.geometry] to calculate the speed
```

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/route2/ReplayRouteMapper.kt [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+        }?.also { points ->
+            replayRouteDriver.drivePointList(options, points)
+                .map { mapToUpdateLocation(it) }
+                .forEach { replayEvents.add(it) }
+        }
+        return replayEvents
+    }
+
+    /**
+     * Use any a polyline string to create replay events that simulator a car driving
```
Comment (Guardiola31337) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/route2/ReplayRouteMapper.kt:66]: Typo? *`that simulator a car driving the route.`

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/route2/ReplayRouteMapper.kt [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+     * @param geometry is a polyline precision 6
+     */
+    fun mapGeometry(geometry: String): List<ReplayEventBase> {
+        return replayRouteDriver.driveGeometry(options, geometry)
+            .map { mapToUpdateLocation(it) }
+    }
+
+    /**
+     * Take a [DirectionsRoute] and map it to events that can be replayed by the [ReplayHistoryPlayer].
+     * Uses the directions api [LegAnnotation] to create the speed and position
```
Comment (Guardiola31337) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/route2/ReplayRouteMapper.kt:78]:
```suggestion
     * Uses the Directions API [LegAnnotation] to create the speed and position
```

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/route2/ReplayRouteMapper.kt [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+     */
+    fun mapDirectionsRouteLegAnnotation(directionsRoute: DirectionsRoute): List<ReplayEventBase> {
+        return directionsRoute.legs()?.flatMap { routeLeg ->
+            mapRouteLegAnnotation(routeLeg)
+        } ?: emptyList()
+    }
+
+    /**
+     * Given a [RouteLeg], use the [LegAnnotation] to create the speed and locations.
+     * To use this, your directions request must include [DirectionsCriteria.ANNOTATION_SPEED] and
```
Comment (Guardiola31337) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/route2/ReplayRouteMapper.kt:91]:
```suggestion
     * To use this, your Directions request must include [DirectionsCriteria.ANNOTATION_SPEED] and
```

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/route2/ReplayRouteDriver.kt [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+        }
+
+        routeInterpolator.createBearingProfile(replayRouteLocations)
+
+        return replayRouteLocations
+    }
+
+    /**
+     * Given a [RouteLeg], use the [LegAnnotation] to create the speed and distance calculations.
+     * To use this, your directions request must include [DirectionsCriteria.ANNOTATION_SPEED] and
```
Comment (Guardiola31337) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/route2/ReplayRouteDriver.kt:61]:
```suggestion
     * To use this, your Directions request must include [DirectionsCriteria.ANNOTATION_SPEED] and
```

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/route2/ReplayRouteMapper.kt [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+    fun mapRouteLegAnnotation(routeLeg: RouteLeg): List<ReplayEventBase> {
+        return replayRouteDriver.driveRouteLeg(routeLeg)
+            .map { mapToUpdateLocation(it) }
+    }
+
+    companion object {
+        private const val REPLAY_ROUTE_ACCURACY_HORIZONTAL = 3.0
+
+        /**
+         * Map a Android location into a replay event.
```
Comment (Guardiola31337) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/route2/ReplayRouteMapper.kt:105]:
```suggestion
         * Map an Android location into a replay event.
```

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/route2/ReplayRouteMapper.kt [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+         *
+         * @param eventTimestamp the eventTimestamp for the replay event
+         * @param location Android location to be replayed
+         * @return an event to that can be replayed
```
Comment (Guardiola31337) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/route2/ReplayRouteMapper.kt:109]: Typo? *`an event to that can be replayed`

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/route2/ReplayRouteMapper.kt [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+                    bearing = if (location.hasBearing()) location.bearing.toDouble() else null,
+                    speed = if (location.hasSpeed()) location.speed.toDouble() else null
+                )
+            )
+            Log.i("ReplayRoute", "ReplayRoute mapToUpdateLocation $values")
+            return values
+        }
+
+        /**
+         * Map a Android location into a replay event.
```
Comment (Guardiola31337) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/route2/ReplayRouteMapper.kt:130]:
```suggestion
         * Map an Android location into a replay event.
```

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/route2/ReplayRouteDriver.kt [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+    private val routeSmoother = ReplayRouteSmoother()
+    private val routeInterpolator = ReplayRouteInterpolator()
+
+    private var timeMillis = 0L
+
+    /**
+     * Given a geometry polyline, simulate the locations needed to drive the geometry.
+     *
+     * @param options allow you to control the driver and car behavior
+     * @param geometry polyline string at precision 6
```
Comment (Guardiola31337) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/route2/ReplayRouteDriver.kt:23]:
```suggestion
     * @param geometry polyline string at [DirectionsCriteria.GEOMETRY_POLYLINE6]
```

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/route2/ReplayRouteDriver.kt [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+
+        for (i in 0 until smoothLocations.lastIndex) {
+            val segmentStart = smoothLocations[i]
+            val segmentEnd = smoothLocations[i + 1]
+            val segmentRoute = routeSmoother.segmentRoute(
+                distinctPoints,
+                segmentStart.routeIndex!!,
+                segmentEnd.routeIndex!!)
+            replayRouteLocations.addInterpolatedLocations(options, segmentRoute, segmentStart, segmentEnd)
+        }
```
Comment (Guardiola31337) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/route2/ReplayRouteDriver.kt:52]: What about extracting this `for` loop into a private function and give them a proper name?

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/route2/ReplayRouteDriver.kt [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+
+        return replayRouteLocations
+    }
+
+    /**
+     * Given a [RouteLeg], use the [LegAnnotation] to create the speed and distance calculations.
+     * To use this, your directions request must include [DirectionsCriteria.ANNOTATION_SPEED] and
+     * [DirectionsCriteria.ANNOTATION_DISTANCE].
+     *
+     * @param routeLeg directions api response of a route leg.
```
Comment (Guardiola31337) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/route2/ReplayRouteDriver.kt:64]:
```suggestion
     * @param routeLeg Directions API response of a [RouteLeg].
```

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/route2/ReplayRouteMapper.kt [hunk 16]
```diff
@@ -0,0 +1,153 @@
...
+        }
+        val geometry = directionsRoute.geometry() ?: return emptyList()
+        return mapGeometry(geometry)
+    }
+
+    /**
+     * Take a single [RouteLeg] from the [DirectionsRoute] and map it to a drive using the
+     * [LegStep.geometry] composed together.
+     *
+     * @param routeLeg the route leg that is converted to replay events
```
Comment (Guardiola31337) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/route2/ReplayRouteMapper.kt:50]:
```suggestion
     * @param routeLeg the [RouteLeg] that is converted to replay events
```

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/route2/ReplayRouteMapper.kt (hunk 6 above)
Comment (Guardiola31337) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/route2/ReplayRouteMapper.kt:66]:
```suggestion
     * Use any polyline string to create replay events that simulator a car driving
```

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/route2/ReplayRouteMapper.kt [hunk 17]
```diff
@@ -0,0 +1,153 @@
...
+                .forEach { replayEvents.add(it) }
+        }
+        return replayEvents
+    }
+
+    /**
+     * Use any a polyline string to create replay events that simulator a car driving
+     * the route.
+     *
+     * @param geometry is a polyline precision 6
```
Comment (Guardiola31337) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/route2/ReplayRouteMapper.kt:69]:
```suggestion
     * @param geometry is a [DirectionsCriteria.GEOMETRY_POLYLINE6]
```

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/route2/ReplayRouteMapper.kt [hunk 18]
```diff
@@ -0,0 +1,153 @@
...
+            mapRouteLegAnnotation(routeLeg)
+        } ?: emptyList()
+    }
+
+    /**
+     * Given a [RouteLeg], use the [LegAnnotation] to create the speed and locations.
+     * To use this, your directions request must include [DirectionsCriteria.ANNOTATION_SPEED] and
+     * [DirectionsCriteria.ANNOTATION_DISTANCE]
+     *
+     * @param routeLeg The route leg to be mapped into replay events
```
Comment (Guardiola31337) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/route2/ReplayRouteMapper.kt:94]:
```suggestion
     * @param routeLeg the [RouteLeg] to be mapped into replay events
```

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/route2/ReplayRouteDriver.kt (hunk 14 above)
Comment (kmadsen) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/route2/ReplayRouteDriver.kt]: yeah iterated some more =)

**Comment by abhishek1508** [comment]
>The UI SDK relies on the ReplayRouteLocationEngine for the simulation mode. As a follow up we should think about how to integrate the new ReplayHistoryLocationEngine and these improvements cc @abhishek1508

I don't think there is any value integrating this to UI SDK 1.0 since we are not going to go public with this. Any new changes or issues on `ReplayRoute` feature in the `core`, would force us to maintain it and fix it in 1.0 private releases as well. I would strongly advocate it to be a part of 2.0 as it would then be a good selling point for the new SDK. I would suggest we keep the `ReplayRouteEngine` as it is for the UI SDK to be able to use it. Once we integrate the new version with 2.0, we can remove `ReplayRouteEngine` from `core`

cc @Guardiola31337 @zugaldia

**Comment by Guardiola31337** [comment]
> I would suggest we keep the `ReplayRouteEngine` as it is for the UI SDK to be able to use it.

"Problem" is that `ReplayRouteLocationEngine` has changed a bit (for the better 😅).

> Once we integrate the new version with 2.0, we can remove `ReplayRouteEngine` from `core`

Would love to remove duplicated code `route2` before releasing `1.0` Core 😬 

BTW don't think it'd be too difficult to migrate (@kmadsen correct if I'm wrong).

**Comment by kmadsen** [comment]
> I would suggest we keep the ReplayRouteEngine as it is for the UI SDK to be able to use it. Once we integrate the new version with 2.0, we can remove ReplayRouteEngine from core

@abhishek1508 the `ReplayRouteEngine` isn't able to handle multiple stops, and it can't stay on the route it's simulating: https://github.com/mapbox/mapbox-navigation-android/issues/2717. 

Like Pablo is saying, sooner we can use the new player, sooner we can delete the old one. But wait for this to land before migrating https://github.com/mapbox/mapbox-navigation-android/pull/2921

**Comment by kmadsen** [comment]
> @abhishek1508 the ReplayRouteEngine isn't able to handle multiple stops, and it can't stay on the route it's simulating: #2717.

I noticed this ticket again and want to follow up. ReplayRoute does support multiple stops. In order to make it work, you need to register the ReplayProgressObserver

**Pull request merged**